File: backend/BudgetToWealthApi/Models/InvestmentReturns/HoldingInvestmentReturn.cs
public class HoldingInvestmentReturn : BaseEntity
{
    public required Guid StartHoldingSnapshotId { get; set; }
    public HoldingSnapshot? StartHoldingSnapshot { get; set; }
    public required Guid EndHoldingSnapshotId { get; set; }
    public HoldingSnapshot? EndHoldingSnapshot { get; set; }
    public required long TotalContributions { get; set; }
    public required long TotalWithdrawals { get; set; }

    public string? UserId { get; set; }

    public decimal? ReturnPercentage
    {
        get
        {
            if (StartHoldingSnapshot == null || EndHoldingSnapshot == null || StartHoldingSnapshot.Balance == 0)
                return null;
            
            var numerator = EndHoldingSnapshot.Balance - TotalContributions - StartHoldingSnapshot.Balance + TotalWithdrawals;
            return decimal.Round((decimal)numerator / (decimal)StartHoldingSnapshot.Balance * 100, 2);
        }
    }
}

File: frontend/app/net-worth/investment-returns/components/form/holding-investment-return-form/HoldingInvestmentReturn.ts
import { HoldingSnapshot } from "../../../../holding-snapshots/components";

export type HoldingInvestmentReturn = {
  id?: number;
  startHoldingSnapshotId: string;
  startHoldingSnapshot?: HoldingSnapshot;
  endHoldingSnapshotId: string;
  endHoldingSnapshot?: HoldingSnapshot;
  totalContributions: number;
  totalWithdrawals: number;
  returnPercentage?: number;
  userId?: string;
}